### ISSUE TITLE
503  Service Unavailable when  canary release with istio traffic manage

### ISSUE THREAD

**Issue by TaylorHere** [argoproj/argo-rollouts#587]
``` yaml
# Source: demo/templates/rollout.yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: demo-rollout-deploy
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 100%
      maxUnavailable: 0
    type: RollingUpdate
  selector:
    matchLabels:
      app: demo
      release: demo
      app.kubernetes.io/name: demo
      app.kubernetes.io/instance: demo
  template:
    metadata:
      annotations:
        checksum/config: 819666c3c37d0b9a43595609f9bf286d3810e9e129472fdf6cb60cbacae6e1d3
      labels:
        app: demo
        release: demo
        app.kubernetes.io/name: demo
        app.kubernetes.io/instance: demo
    spec:
      imagePullSecrets:
        - name: lilith-image
      serviceAccountName: demo
      containers:
        - name: demo
          env:
            - name: LC_ALL
              value: en_US.utf-8
            - name: LANG
              value: en_US.utf-8
          securityContext:
            {}
          image: "demo:v1"
          imagePullPolicy: IfNotPresent
          ports:
            - name: http
              containerPort: 5000
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /api/live
              port: 5000
          readinessProbe:
            httpGet:
              path: /api/ready
              port: 5000
          resources:
            limits:
              cpu: "1"
              memory: 256Mi
            requests:
              cpu: "1"
              memory: 256Mi
          volumeMounts:
            - mountPath: /app/anti_fraud_server/configs/
              name: demo
      volumes:
        - name: demo
          configMap:
              name: demo
  minReadySeconds: 10
  revisionHistoryLimit: 3
  progressDeadlineSeconds: 600
  strategy:
    canary:
      canaryService: demo-canary-svc
      stableService: demo-prod-svc
      maxUnavailable: 0
      maxSurge: 25%
      steps:
      - pause: {}
      - setWeight: 0
      - pause: {}
      - setWeight: 50
      - pause: {}
      - setWeight: 100
      trafficRouting:
        istio:
          virtualService: 
            name: demo
            routes:
            - demo-vpc
            - demo-public-debug

```

when I apply my rollout to k8s, a New ReplicaSet created, with istio vsvc traffic 100% to demo-prod-svc

503 Service Unavailable occurred for about 1s in client side.

in this state, no new pods created, demo-prod-svc has no changes.

**Comment by TaylorHere** [comment]
when canary setWeight as 100%, canary become stable, 503 Service Unavailable occurred again.

**Comment by TaylorHere** [comment]
Rollout Created with step: pause, 503 occurred.
![image](https://user-images.githubusercontent.com/12156543/87540815-d956d600-c6d2-11ea-85f1-eb5cd852debc.png)


In progress.
![image](https://user-images.githubusercontent.com/12156543/87540119-af50e400-c6d1-11ea-995f-6263b692e153.png)

Rollout canary switch to stable, 503 occurred.

![image](https://user-images.githubusercontent.com/12156543/87541218-7dd91800-c6d3-11ea-8724-52f3fe030da4.png)

**Comment by jessesuen** [comment]
FYI, you have spec.strategy listed twice. I think it is still working because the latter spec.strategy is overwriting, but I would not rely on that. 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: demo-rollout-deploy
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 100%
      maxUnavailable: 0
    type: RollingUpdate
```

Could you clarify the behavior of 503? Do they only happen when the weight is adjusted? And only last a small duration (1 second) and then reach a steady/reliable state?

**Comment by TaylorHere** [comment]
503 occurred at the rollout created immediately, and the last step which is setWeight: 100
I'll do more test on this latter.

**Comment by TaylorHere** [comment]
another odd thing, when the cluster receive a large traffic, like 18k/s, all the rollout deploy and ReplicaSet vanished, all the pods been deleted.

**Comment by TaylorHere** [comment]
and I noticed that, my traffic goes to prod and canary at the same time, even if my prod weight is 100

**Comment by nemo83** [comment]
Couple of questions. When you say 
> Rollout Created with step: pause, 503 occurred.
Does it mean you're creating the Rollout for the first time, ie to run alongside a deployment you will then decommission?
Or are you _just_ deploying a new version?

Can you also please paste the `Service` definitions please? Wondering if there is any funny label there that is messing around.

**Comment by jessesuen** [comment]
> another odd thing, when the cluster receive a large traffic, like 18k/s, all the rollout deploy and ReplicaSet vanished, all the pods been deleted.

I'm certain this is not caused by the rollout-controller. It does not have RBAC privileges to delete a rollout object.

Also, did you fix this YAML bug that I pointed out earlier?

```yaml
  strategy:
    rollingUpdate:
      maxSurge: 100%
      maxUnavailable: 0
    type: RollingUpdate
```

**Comment by BEvgeniyS** [comment]
I can confirm that this issue exists.

We run a slowcooking type of the load test which continuously queries our service.  
We're using EKS 1.14, istio 1.4.6 and argo-rollouts 0.8.2.  

We were able to repeatedly  introduce 503 errors by just having a single `pause` as our canary strategy.
As soon as rollouts manifest is edited (we just "tickled" it by changing resource limit in spec) and argo rollouts controller starts working, we observe a small portion of errors (we ran about 10 requests per second, we got ~20-30 errors).

So no new pods are created, rollouts is waiting for us to manually promote it to next step, but errors already happened.

This is easily reproducable in our cluster, and the issue goes away if we:  
a) Switch back to deployments
b) Use primitive canary strategy without istio virtualservice weights.  

It appears that there's an issue with how rollout is working with service/vs/endpoints/routes.  

Please let us know if you need any additional data or if we can somehow help you with this issue otherwise.

**Comment by jessesuen** [comment]
> We were able to repeatedly introduce 503 errors by just having a single pause as our canary strategy.

@BEvgeniyS we are aware of an issue where the rollout controller may be scaling down the old replicaset faster than istio can shift weight to the new version. Could you attempt this canary strategy:

```yaml
steps:
- setWeight: 100
- pause:
    duration: 1m
```

The idea behind the 1m pause, is that it is more than enough time for the change in weights to get propagated to istio sidecars.

**Comment by BEvgeniyS** [comment]
@jessesuen the strategies we tried initially were

```
steps:
- setWeight: 20
- pause:
    duration: 1m
- setWeight: 40
- pause:
    duration: 1m
....
```
and
```
```
steps:
- setWeight: 0
- pause:
    duration: 1m
- setWeight: 20
- pause:
    duration: 1m
....
```

And this was what initially brought us here.  
Does it makes sense to test it with setWeight: 100? 
```

**Comment by nemo83** [comment]
> @jessesuen the strategies we tried initially were
> 
> ```
> steps:
> - setWeight: 20
> - pause:
>     duration: 1m
> - setWeight: 40
> - pause:
>     duration: 1m
> ....
> ```
> 
> and
> 
> ```
> steps:
> - setWeight: 0
> - pause:
>     duration: 1m
> - setWeight: 20
> - pause:
>     duration: 1m
> ....
> ```
> 
> And this was what initially brought us here.
> Does it makes sense to test it with setWeight: 100? Because seems to me that will just route ALL the traffic to the canary 

**TL;DR** yeah it does make sense.

## Long explanation

service which has 0 pods, and of course every request will fail
```steps:
- setWeight: 0
- pause:
    duration: 1m```
can be ditched, as it has no effect and provides no information.

The underlying issue here is that the service does not handle SIGTERM, or it handles but cannot complete their in-flights requests on time (usually 30 seconds), that is nothing to do w/ Argo in neither the cases.

The one below is a workaround that will allow OLD pods to stop receiving traffic (as 100% will go to canary) and then have 1 full minute to complete in-flight transactions.
```
- setWeight: 100
- pause:
    duration: 1m
```

**Comment by BEvgeniyS** [comment]
@nemo83 I will test this tomorrow,  however it still not clear to me how the service  is supposed to handle sigterm. Istio's envoy is draining connections for a minute. If we set the canary strategy to just pause without changing any weights we still observe the errors. No pods created/destroyed. 

Sorry but your explanation  doesn't make sense to me

**Comment by nemo83** [comment]
> @nemo83 I will test this tomorrow, however it still not clear to me how the service is supposed to handle sigterm. Istio's envoy is draining connections for a minute.

Not exactly, if you set 40% weight, the other 60% is still going to the old pods, and istio is not _draining_ anything. Just sending less traffic.

> 
> If we set the canary strategy to just pause without changing any weights we still observe the errors. No pods created/destroyed.

I am not clear at what moment in time you're seeing the errors, but if this is any similar to the issue I was experiencing, 503 are thrown at the last step (that in your case is implicit).

Your spec is:

```
steps:
- setWeight: 0
- pause:
    duration: 1m
- setWeight: 20
- pause:
    duration: 1m
```

What really argo runs is:

```
steps:
- setWeight: 0
- pause:
    duration: 1m
- setWeight: 20
- pause:
    duration: 1m
- setWeight: 100
```
Once the 1 minute wait expires, the setWeight: 100% is implicitly executed and 100% of *new* traffic is sent to canary, and the old pods receive a SIGTERM. (so in this case I think your assumption of _istio draining for 1 minute and no pods are destroyed is wrong_)
Now, new requests are going to canary (now stable) pods, BUT there are still in-flight transactions to OLD pods, that have just received SIGTERM. If SIGTERM was managed properly, old pods won't die right away, but would still be running for a little bit, enough to serve in-flight transactions, then die. If SIGTERM is un-handled pods will just die and in-flight transactions cause the error.
The additional wait is an explicit man-made grace period to allow old pods to finish their work before even being killed.
 
> Sorry but your explanation doesn't make sense to me

Of course I'm not 100% sure this is the same issue I had back in the days, and I was confused myself :D . But if it is, it will all make sense in the end.

**Comment by BEvgeniyS** [comment]
Here's the strategy I was talking about:

```
  strategy:
    canary:
      maxUnavailable: 0
      maxSurge: 3
      steps:
      - pause: {}
      canaryService: myapp-canary
      stableService: myapp
      trafficRouting:
        istio:
           virtualService:
            name: myapp-services
            routes:
            - myapp-routes
```

Errors happen as soon as the only step comes into effect (shortly after applying the manifest/editing the rollouts objects)

**Comment by BEvgeniyS** [comment]
> Not exactly, if you set 40% weight, the other 60% is still going to the old pods, and istio is not _draining_ anything. Just sending less traffic.

https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig
It does drain. There's another issue with the sidecar/app container dependencies, but this is all irrelevant. No new pods created, no old pods killed, errors observed

**Comment by nemo83** [comment]
Okay, I should have read the full thread. This is in fact a different issue from what I had in mind.

I have a few questions.

1. do you have also a deployment running in parallel?
2. are you applying the `Rollout` manifest for the first time, or is it already there? (I'm sure it's already there and you're amending it... but I just wanna be sure)
3. would you mind posting the two vanilla kubernetes services spec and the Istio Virtual Service?

I had a problem in the past where, when migrating an app from vanilla Kubernetes Deployment to Rollout, services labels where changed in real time and messing up w/ routing.

**Comment by BEvgeniyS** [comment]
Thanks @nemo83,

1. No, we just applied the rollouts manifest.
2. It was already there =)

3. services.yaml:
```yaml
# Stable service
apiVersion: v1
kind: Service
metadata:
  name: myapp
  labels:
    app: myapp
    service: myapp
spec:
  ports:
  - port: 80
    targetPort: 8080
    protocol: TCP
    name: http
  selector:
    app: myapp
  type: ClusterIP

---
# Canary service
apiVersion: v1
kind: Service
metadata:
  name: myapp-canary
  labels:
    app: myapp
    service: myapp-canary
spec:
  ports:
  - port: 80
    targetPort: 8080
    protocol: TCP
    name: http
  selector:
    app: myapp
  type: ClusterIP
```

virtualservice.yaml:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: myapp-services
  namespace: myns
spec:
  hosts:
  - "*"
  exportTo:
  - "*"
  gateways:
  - istio-ingress.myns.svc.cluster.local
  - mesh
  http:
  - name: myapp-routes
    match:
    - uri:
        prefix: /
    route:
      - destination:
          host: myapp
          port:
            number: 80
        weight: 100
      - destination:
          host: myapp-canary
          port:
            number: 80
        weight: 0
```

rollouts.yaml:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: myapp
  labels:
    app: myapp
    app.kubernetes.io/name: myapp
spec:
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app.kubernetes.io/name: myapp
        app: myapp
    spec:
      serviceAccountName: myapp
      containers:
        - name: myapp
          image: myapp-image
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /healthz
              port: 8080
            initialDelaySeconds: 60
            periodSeconds: 10
            failureThreshold: 20
            successThreshold: 1
          readinessProbe:
            httpGet:
              path: /health
              port: 8080
            failureThreshold: 2
            successThreshold: 6
            initialDelaySeconds: 10
            periodSeconds: 10
  strategy:
    canary:
      maxUnavailable: 0
      maxSurge: 3
      steps:
      - pause: {}
      canaryService: myapp-canary
      stableService: myapp
      trafficRouting:
        istio:
           virtualService:
            name: myapp-services
            routes:
            - myapp-routes
```


Yes, migrating from deployment to rollouts was a different story :)

**Comment by nemo83** [comment]
@BEvgeniyS okay, this looks good to me.

How do you trigger the deploy?

These are manifests for a test app I'm running:

```
---
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: rollout-canary
spec:
  replicas: 2
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: rollout-canary
  template:
    metadata:
      labels:
        app: rollout-canary
    spec:
      containers:
        - name: rollouts-demo
          image: {{ .Values.image }}
          imagePullPolicy: Always
          ports:
            - containerPort: 8080
  strategy:
    canary:
      steps:
        - setWeight: 50
        - pause:
            duration: 30
      canaryService: rollout-canary # required
      stableService: rollout-canary-stable  # required
      trafficRouting:
        istio:
          virtualService:
            name: rollout-canary-vs  # required
            routes:
              - primary # At least one route is required
```

```
---
apiVersion: v1
kind: Service
metadata:
  name: rollout-canary-stable
spec:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8080
  selector:
    app: rollout-canary
---
apiVersion: v1
kind: Service
metadata:
  name: rollout-canary
spec:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8080
  selector:
    app: rollout-canary
```

```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: rollout-canary-vs
  labels:
    app.kubernetes.io/name: applications.argoproj.io
    app.kubernetes.io/part-of: argocd
    app.kubernetes.io/instance: argocd
    app.kubernetes.io/managed-by: Tiller
spec:
  hosts:
  - rollout-canary.flo.pub
  gateways:
    - rollout-canary-flow-pub
  http:
    - name: primary
      route:
        - destination:
            host: rollout-canary-stable
          weight: 100
        - destination:
            host: rollout-canary
          weight: 0
```

**Comment by pliutak-nih** [comment]
We have the same issue.
I raised a bug for Istio team reflecting the actions Argo Rollouts took during canary deployment https://github.com/istio/istio/issues/26861.

**Comment by BEvgeniyS** [comment]
@nemo83 as noted above, doesn't matter how I trigger it. We tried direct object editing and helm deployment - result is the same

**Comment by hari2192** [comment]
Following up on issue reported by @BEvgeniyS, we still facing this issue:

We have tried to update our canary strategy by adding extra step to at the end:
`steps:
      - setWeight: 20
      - pause:
          duration: 5m
      - setWeight: 40
      - pause:
          duration: 5m
      - setWeight: 60
      - pause:
          duration: 5m
      - setWeight: 80
      - pause:
          duration: 5m
      - setWeight: 100
      - pause:
          duration: 1m`

As soon as the last step is completed i.e set weight 100% and pause for 1 min, we started seeing bunch of `503 NR` errors from old replica set (which meant to not receive any traffic), so we have suspicion that traffic is still being routed to old replica set.

Any help to explain this behavioiur ?

**Comment by jessesuen** [comment]
Hello, I'm just catching up on this issue so I apologize if I missed something. So the original suggestion of:

```yaml
      - setWeight: 100
      - pause:
          duration: 1m
```

Was to accomplish two things:

1. Allow time for Istio itself, to affect the change in weights *before* Argo Rollouts decides to scale down the old pods. Currently, as soon as the last step is completed, the rollout controller immediately starts scaling down the old ReplicaSet, which I feel is premature. Issue https://github.com/argoproj/argo-rollouts/issues/478 is to have implicit scaleDownDelay when trafficRouting is used. In other words the implicit sequence would be:
    * reach the end of the steps
    * adjust weight to 100
    * wait 30 seconds
    * scale down old replicaset

2. The second reason for the 1 minute delay after setWeight: 100 was this: For applications which do not gracefully shutdown as a result of a SIGTERM, the one minute delay would give sufficient time for the old ReplicaSet to stop receiving new connections, before receiving a request to shutdown. This would allow ReplicaSet to drain connections naturally

The big stick in these two theories, is that the 503s are able to be reproduced without any SIGTERMs involved at all. Simply the act of patching a service, is able to reproduce the problem.

Based on the recent comments in https://github.com/istio/istio/issues/26861, it appears it could be an upstream istio issue. But will monitor the issue to see if the rollout behavior can/should be adjusted to prevent the problem.

**Comment by jessesuen** [comment]
I think a new way of achieving canary in Argo Rollouts to allow a single service but leverage a DestinationRule with canary/stable subsets, would help avoid this issue. The feature enhancement is being discussed here:

https://github.com/argoproj/argo-rollouts/issues/617

**Comment by Eslamanwar** [comment]
I can confirm that we are facing the same issue with istio/Argo rollout on k8s ,the 503 happens exactly when at the start and the end of the canary route shifting . any workaround we can follow to mitigate this issue till we have DestinationRule with canary/stable subsets

**Comment by niscoveanumircea** [comment]
I confirm that this is also happening with :
```
Istio client version: 1.7.4
Istio control plane version: 1.7.4
Istio data plane version: 1.7.4 (72 proxies)
Kubernetes: v1.19.4
argocd: v1.7.8+ef5010c
kubectl-argo-rollouts: v0.10.0+3837eab
```

![image](https://user-images.githubusercontent.com/22934144/99911086-c09ea280-2cfa-11eb-8a6e-b6b46ce09b79.png)

When I first noticed the 503 NR responses, these kind of responses appeared both when the rollout started and when the rollout ended (right before the termination of the old pods).

Using the following strategy I thought I am going to prevent the apparition of the 503 codes when the rollout started  using `setCanaryScale: { replicas: 1 }`  right before redirecting traffic to the canary service. But actually, the 503 codes were coming from the 'stable-service' and not the 'canary-service'

`
  strategy:
    canary: #Indicates that the rollout should use the Canary strategy
      maxSurge: "100%"
      maxUnavailable: 0 # The maximum number of pods that can be unavailable during the update
      steps:
      - setWeight: 0
      - setCanaryScale: { replicas: 1 }
      - pause: { duration: 30s }
      - setWeight: 20
      - pause: { duration: 30s }
      - setCanaryScale:
          matchTrafficWeight: true
      - setWeight: 30
      - pause: { duration: 30s }
      - setWeight: 50
      - pause: { duration: 30s }
      - setWeight: 100
      - pause: { duration: 30s }
`
I am still facing issues with 503 responses when the rollout is starting - at step 1 ( before any traffic redirection ). I also reproduced with the below strategy. I got 4 x 503 responses while sending traffic via fortio `fortio load -qps 200 -t 0s -H "Host: reconciliation-api" reconciliation-api/api/v1/openapi.json`
`
  strategy:
    canary: #Indicates that the rollout should use the Canary strategy
      maxSurge: "100%"   # Defines the maximum number of replicas the rollout can create to move to the correct ratio set by the last setWeight
      maxUnavailable: 0 # The maximum number of pods that can be unavailable during the update
      steps:
      - setWeight: 0
      - setCanaryScale: { replicas: 1 }
      - pause: {} # pause indefinitely
`

**Comment by niscoveanumircea** [comment]
The previously described behavior is not reproducing with Istio 1.8.0 !

**Comment by romilpunetha** [comment]
I'm still facing this issue using istio traffic routing using destination rule on istio 1.12.

The problem is that after the weight is shifted 100% to stable subset at the end of the canary process, the caller service does not have endpoints for the stable endpoint cluster `outbound|80|stable|x.svc.cluster.local`. The canary pods are still tracked under the canary endpoint cluster. The flip only happens when the pods enter the terminating stage. 

<img width="1592" alt="Screenshot 2022-01-04 at 3 58 39 AM" src="https://user-images.githubusercontent.com/47735999/148005245-2861ea81-de0a-429c-9ff2-ebd05715876c.png">


<img width="1649" alt="Screenshot 2022-01-04 at 3 58 06 AM" src="https://user-images.githubusercontent.com/47735999/148005236-73184ef6-834a-483b-bb30-240eaf68dc1e.png">
